source/fundamentals/data-formats/pojo-customization.txt

```diff
@@ -10,78 +10,653 @@ POJO Customization
+
+
+.. _conventions:
 
 Conventions
 -----------
 
-- Description
-- Table of preset Conventions
-  - Link to Annotations section
-- Example on how to specify a Convention
-- How to create a custom Convention
+The ``Convention`` interface contains configuration options that modify the
+behavior of a ``ClassModel`` or ``PropertyModel``. You can specify a
+``Convention`` in a call to ``PojoCodecProvider.Builder.conventions()`` or to
+``ClassModelBuilder.conventions()``. The builders apply ``Convention``
+instances in order which may override behavior defined in one applied earlier.
+
+You can access the ``Convention`` instances defined in the BSON library
+from the following static fields in the ``Conventions`` class:
+
+.. list-table::
+   :header-rows: 1
+   :stub-columns: 1
+
+   * - Field Name
+     - Description
+
+   * - ``ANNOTATION_CONVENTION``
+     - Enables the annotations defined in the ``org.bson.codecs.pojo.annotations``
+       package for your POJO.
+
+   * - ``CLASS_AND_PROPERTY_CONVENTION``
+     - | Sets the following default values for the ``ClassModel`` and ``PropertyModel`` instances:
+       | - Discriminator key to ``_t``
+       | - Discriminator value to the ``ClassModel`` simple type name
+       | - Id field to ``_id`` for each ``PropertyModel``.
+
+   * - ``DEFAULT_CONVENTIONS``
+     - | Enables the following Conventions:
+       | - ``CLASS_AND_PROPERTY_CONVENTION``
+       | - ``ANNOTATION_CONVENTION``
+       | - ``OBJECT_ID_GENERATORS``
+
+   * - ``NO_CONVENTIONS``
+     - Provides an empty list.
+
+   * - ``OBJECT_ID_GENERATORS``
+     - Adds a default ``IdGenerator`` that adds a new ``ObjectId`` for each
+       ``ClassModel`` that use ``ObjectId`` values in the ``id`` property.
+
+   * - ``SET_PRIVATE_FIELDS_CONVENTION``
+     - Enables the ``ClassModel`` to set private fields using reflection
+       without requiring a setter method.
+
+   * - ``USE_GETTERS_FOR_SETTERS``
+     - Enables use of getter methods as setters for ``Collection`` and ``Map``
+       fields if no setter method exists.
+
+You can specify Conventions using one of the following methods:
+
+- :java-docs:`PojoCodecProvider.Builder.conventions() </apidocs/bson/org/bson/codecs/pojo/PojoCodecProvider.Builder.html>`
+- :java-docs:`ClassModelBuilder.conventions() </apidocs/bson/org/bson/codecs/pojo/ClassModelBuilder.html#conventions(java.util.List)>`
+
+To create a custom Convention, create a class that implements the
+``Convention`` interface and override the ``apply()`` method from which you can
+access your ``ClassModelBuilder`` instance.
+
+.. _annotations:
 
 Annotations
 -----------
 
-- Description
-- Table of Annotations
-- Example class with annotations
-- Example output
+You can apply annotations to the getter and setter methods of a POJO
+class. These annotations configure the ``ClassModel`` and ``PropertyModel``
+behavior for a specific field, method, or class.
```

Comment:
It might be helpful to mention how conflicts between annotations and Conventions are resolved. I assume that annotations override conventions if annotations are enabled, but it might be nice to have an explicit callout of that fact (or, if there are exceptions to that behavior, to warn users).

Reply:
Thanks, I'll link to the next section.

Reply:
Not sure what you mean, but maybe the sentence that explains that conventions are applied in order https://github.com/mongodb/docs-java/pull/52/files#diff-81770bd4e69d34b6329eacade6c66cffd45edab4a1eb84310dc644364e1660deR188-R189 needs to be more prominient. If the ANNOTATION_CONVENTION convention is not applied, the annotations do not get applied. If the convention is applied, it overrides whatever was specified earlier.

Reply:
Perhaps I wasn't clear here -- i meant situations where ANNOTATION_CONVENTION is enabled, and a specific annotation instance specifies something that conflicts with another convention (for instance, a situation where a user specifies a custom discriminator field/value but has the default discriminator convention enabled as well. I assume that the annotation for a custom discriminator field/value would "win" in that case, but it could be helpful to specify that here in a sentence.

Reply:
Hm, I think there might be some confusion and I want to clarify to make sure I'm not explaining something incorrectly before moving to external review.

First, could you clarify whether you're talking about:
1) "default values" - something specified only if nothing previously exists
2)  "DEFAULT_CONVENTIONS" - enables three pre-defined Conventions
3) Nothing to do with defaults - different behavior from two or more different Conventions in which case the latest one in the list overrides any conflicting value

I think i was explaining #3 in my prior comment in this thread.

For #1, suppose we are talking about CLASS_AND_PROPERTY_CONVENTION which provides those (description says "Sets the following default values..."). These defaults only apply if no other value is specified at that time. And, if any discriminator key/value is added to the class via PojoCodecProvider.Builder.conventions() or ClassModelBuilder.conventions() after the CLASS_AND_PROPERTY_CONVENTION is specified, it overwrites that value. If I first defined CLASS_AND_PROPERTY_CONVENTION, my "Product" class would have a "default" discriminator key "_t" and discriminator "Product". If my annotation specified @BsonDiscriminator(key="_cls", value="AnnotatedProduct"), that takes precedence whether CLASS_AND_PROPERTY_CONVENTION is defined before or after the ANNOTATION_CONVENTION.

For #2, I don't think any of the three conventions conflict. You can have situations such as #1 (in which an annotation overrides the defaults provided in CLASS_AND_PROPERTY_CONVENTION) or #3 (in which additional Conventions are specified to PojoCodecProvider.Builder.conventions() or ClassModelBuilder.conventions()).

Let me know if one of these answers your questions. Happy to chat over Zoom if that helps us get to the issue more quickly.

Reply:
Please ignore links generated by github markdown to "#1", "#2", and "#3"

Reply:
#1 answered my question -- specifically, "If my annotation specified @BsonDiscriminator(key="_cls", value="AnnotatedProduct"), that takes precedence whether CLASS_AND_PROPERTY_CONVENTION is defined before or after the ANNOTATION_CONVENTION." I was just suggesting that we could add a sentence explicitly calling out that relationship (between @BsonDiscriminator and CLASS_AND_PROPERTY_CONVENTION) but it might be that I missed a previous callout.